File: src/main/java/me/testaccount666/serversystem/commands/executables/offlineteleport/CommandOfflineTeleport.java
package me.testaccount666.serversystem.commands.executables.offlineteleport;

import me.testaccount666.serversystem.ServerSystem;
import me.testaccount666.serversystem.commands.ServerSystemCommand;
import me.testaccount666.serversystem.commands.executables.AbstractServerSystemCommand;
import me.testaccount666.serversystem.managers.PermissionManager;
import me.testaccount666.serversystem.userdata.CachedUser;
import me.testaccount666.serversystem.userdata.ConsoleUser;
import me.testaccount666.serversystem.userdata.User;
import org.bukkit.command.Command;
import org.bukkit.entity.Player;

import java.util.Optional;

import static me.testaccount666.serversystem.utils.MessageBuilder.command;
import static me.testaccount666.serversystem.utils.MessageBuilder.general;

@ServerSystemCommand(name = "offlineteleport", variants = "offlineteleporthere", tabCompleter = TabCompleterOfflineTeleport.class)
public class CommandOfflineTeleport extends AbstractServerSystemCommand {

    @Override
    public void execute(User commandSender, Command command, String label, String... arguments) {
        if (commandSender instanceof ConsoleUser) {
            general("NotPlayer", commandSender).build();
            return;
        }

        if (command.getName().equalsIgnoreCase("offlineteleport")) {
            handleOfflineTeleport(commandSender, command, label, arguments);
            return;
        }

        if (command.getName().equalsIgnoreCase("offlineteleporthere")) handleOfflineTeleportHere(commandSender, command, label, arguments);
    }

    private void handleOfflineTeleport(User commandSender, Command command, String label, String... arguments) {
        var cachedUserOptional = getTargetUser(commandSender, label, arguments[0]);
        if (cachedUserOptional.isEmpty()) return;

        var cachedUser = cachedUserOptional.get();
        var targetName = cachedUser.getOfflineUser().getName().orElse(arguments[0]);

        commandSender.getPlayer().teleport(cachedUser.getOfflineUser().getLogoutPosition());

        command("OfflineTeleport.Success", commandSender).target(targetName).build();
    }

    private void handleOfflineTeleportHere(User commandSender, Command command, String label, String... arguments) {
        var cachedUserOptional = getTargetUser(commandSender, label, arguments[0]);
        if (cachedUserOptional.isEmpty()) return;

        var cachedUser = cachedUserOptional.get();
        var targetName = cachedUser.getOfflineUser().getName().orElse(arguments[0]);

        cachedUser.getOfflineUser().setLogoutPosition(commandSender.getPlayer().getLocation());
        cachedUser.getOfflineUser().save();

        command("OfflineTeleportHere.Success", commandSender).target(targetName).build();
    }

    private Optional<CachedUser> getTargetUser(User commandSender, String label, String name) {
        var cachedUserOptional = ServerSystem.Instance.getUserManager().getUser(name);

        if (cachedUserOptional.isEmpty()) {
            general("ErrorOccurred", commandSender).label(label).target(name).build();
            return Optional.empty();
        }

        var cachedUser = cachedUserOptional.get();
        var targetName = cachedUser.getOfflineUser().getName().orElse(name);

        if (cachedUser.isOnlineUser()) {
            general("Offline.NotOffline", commandSender).target(targetName).build();
            return Optional.empty();
        }

        return Optional.of(cachedUser);
    }

    @Override
    public boolean hasCommandAccess(Player player, Command command) {
        if (command.getName().equalsIgnoreCase("offlineteleport"))
            return PermissionManager.hasCommandPermission(player, "OfflineTeleport.Use", false);

        return PermissionManager.hasCommandPermission(player, "OfflineTeleportHere.Use", false);
    }
}

File: src/main/kotlin/me/testaccount666/serversystem/commands/executables/offlineteleport/CommandOfflineTeleport.kt
package me.testaccount666.serversystem.commands.executables.offlineteleport

import me.testaccount666.serversystem.commands.ServerSystemCommand
import me.testaccount666.serversystem.commands.executables.AbstractServerSystemCommand
import me.testaccount666.serversystem.extensions.*
import me.testaccount666.serversystem.userdata.*
import org.bukkit.command.Command

@ServerSystemCommand("offlineteleport", ["offlineteleporthere"], TabCompleterOfflineTeleport::class)
class CommandOfflineTeleport : AbstractServerSystemCommand() {
    override fun minRequiredArguments(command: Command) = 1
    override fun getUsagePermission(command: Command): String {
        if (command.name.equals("offlineteleport", true)) return "OfflineTeleport.Use"
        return "OfflineTeleportHere.Use"
    }

    override fun getSyntaxPath(command: Command?) = "OfflineTeleport"

    override fun execute(commandSender: User, command: Command, label: String, vararg arguments: String) {
        if (!isPlayer(commandSender)) return

        if (command.name.equals("offlineteleport", true)) {
            handleOfflineTeleport(commandSender, label, *arguments)
            return
        }

        handleOfflineTeleportHere(commandSender, label, *arguments)
    }

    private fun handleOfflineTeleport(commandSender: User, label: String, vararg arguments: String) {
        val cachedUser = getTargetUser(commandSender, label, arguments[0]) ?: return
        val targetName = cachedUser.offlineUser.getNameOrNull() ?: arguments[0]

        commandSender.getPlayer()!!.teleport(cachedUser.offlineUser.logoutPosition!!)

        commandSender.commandMsg("OfflineTeleport.Success") { target(targetName) }
    }

    private fun handleOfflineTeleportHere(commandSender: User, label: String, vararg arguments: String) {
        val cachedUser = getTargetUser(commandSender, label, arguments[0]) ?: return
        val targetName = cachedUser.offlineUser.getNameOrNull() ?: arguments[0]

        cachedUser.offlineUser.logoutPosition = commandSender.getPlayer()!!.location
        cachedUser.offlineUser.save()

        commandSender.commandMsg("OfflineTeleportHere.Success") { target(targetName) }
    }

    private fun getTargetUser(commandSender: User, label: String, name: String): CachedUser? {
        val cachedUser = getService<UserManager>().getUserOrNull(name) ?: run {
            commandSender.generalMsg("ErrorOccurred") {
                label(label)
                target(name)
            }
            return null
        }

        val targetName = cachedUser.offlineUser.getNameOrNull() ?: name

        if (cachedUser.isOnlineUser) {
            commandSender.generalMsg("Offline.NotOffline") { target(targetName) }
            return null
        }

        return cachedUser
    }
}
